### PR TITLE
[CDEC-498] Remove `RecordWildCards` extension from cabal default extensions for packages `auxx`, `binary`, `chain`, `util`, `networking`, `crypto`, `core`, `db`, `infra`, `client` and `explorer`

### PR DESCRIPTION
**Pull request by Jimbo4350**
## Description

The `RecordWildCards` extension makes it very difficult to discern where values come from. This is further obfuscated when the extension is not declared at the top of the module and only exists as a default extension in the cabal file. Therefore this PR is one of two PRs that removes the `RecordWildCards` extension from the cabal file and creates a doc (RecordWildCards.md) with a table indicating the type, which file the type's `RecordWildCard` is being used in and which file the type definition exists in.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-498

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/Pos/Core/Common/Address.hs [hunk 1]
```diff
@@ -390,8 +392,8 @@ isRedeemAddress Address {..} =
 isUnknownAddressType :: Address -> Bool
 isUnknownAddressType Address {..} =
     case addrType of
-        ATUnknown {} -> True
-        _            -> False
+        ATUnknown _ -> True
+        _           -> False
```
Comment (erikd) [inline core/src/Pos/Core/Common/Address.hs]: As above

File: infra/src/Pos/Infra/Communication/Protocol.hs [hunk 2]
```diff
@@ -141,14 +141,14 @@ data SpecError
 instance Exception SpecError
 
 instance Buildable SpecError where
-    build (OutSpecNotReported outSpecs spec) =
+    build (OutSpecNotReported outSpecs' spec) =
```
Comment (erikd) [inline infra/src/Pos/Infra/Communication/Protocol.hs]: Why was that needed? Probably best not to change things that don't *need* to be changed.

File: infra/src/Pos/Infra/Network/DnsDomains.hs [hunk 3]
```diff
@@ -153,8 +153,8 @@ resolveDnsDomains :: forall a e.
                   -> Word16                      -- ^ Default port
                   -> [NodeAddr a]
                   -> IO [Either e [(ByteString, Word16)]]
-resolveDnsDomains resolve defaultPort dnsDomains =
-    forM dnsDomains resolveOne
+resolveDnsDomains resolve defaultPort dnsDomains' =
+    forM dnsDomains' resolveOne
```
Comment (erikd) [inline infra/src/Pos/Infra/Network/DnsDomains.hs]: Here I can see a name clash. :+1:

File: core/src/Pos/Core/Common/AddrSpendingData.hs [hunk 4]
```diff
@@ -130,9 +130,9 @@ instance Bi AddrType where
 addrSpendingDataToType :: AddrSpendingData -> AddrType
 addrSpendingDataToType =
     \case
-        PubKeyASD {} -> ATPubKey
-        ScriptASD {} -> ATScript
-        RedeemASD {} -> ATRedeem
+        PubKeyASD _ -> ATPubKey
+        ScriptASD _ -> ATScript
+        RedeemASD _ -> ATRedeem
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply core/src/Pos/Core/Common/AddrSpendingData.hs]: The rationale here was to avoid having an unnecessary `RecordWildCards` extension in the file. I overlooked the difference between  `{}` and `_` so I will revert!

File: core/src/Pos/Core/Common/Address.hs (hunk 1 above)
Comment (Jimbo4350) [reply core/src/Pos/Core/Common/Address.hs]: Will revert here also
